design/20230726-cel-policy.md

```diff
@@ -0,0 +1,368 @@
+- validate CEL expressions in CertificateRequestPolicy on admission
+- extensions to the CertificateRequestPolicy API should be backwards compatible
+
+### Non-Goals
+
+<!--
+What is out of scope for this proposal? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- CEL support in CertificateRequestPolicy `constraints`
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+What is the desired outcome and how do we measure success?
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation- those should go into "Design Details" below.
+-->
+
+The current CertificateRequestPolicy API allows a policy author to specify `allowed` CSR attributes in the
+CertificateRequest, and will deny the request if it contains an attribute which is not present in the `allowed` block.
+To configure valid CSR attribute values, the API provides fields for allowed
+`values` (array; any element matching the attribute is sufficient),
+`value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the request attribute.
+Allowed string fields accept wildcards "*" within its value(s).
+
+This proposal suggests to add support for specifying allowed string fields using CEL expressions, similar to the existing
+wildcard support. While CEL expressions could eventually replace the current CSR attribute validation API - including the 
+wildcard support, the main improvement would be support for policies that specifies an allowed attribute value that depends
+on CertificateRequest fields e.g. `namespace`. If implemented, this new feature would close
+[#62](https://github.com/cert-manager/approver-policy/issues/62).
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this proposal gets implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### X.509 DNS SAN in namespace domain
+
+As a cluster admin in a multi-tenant cluster, I want to provide a cert-manager ClusterIssuer to our users allowing
+self-provisioning of server certificates. Since the cluster is multi-tenant, I want to ensure that tenants do not
+"tip on each other's toes" and enforce that the issuer only signs certificates where the X.509 DNS SAN attribute
+ends with `<namespace>.<domain>`.
+
+A variant of this user story is exemplified in [#62](https://github.com/cert-manager/approver-policy/issues/62),
+and I think it's one of the most common user stories the proposed feature will solve.
+
+#### SPIFFE ID identifying namespace
+
+As a cluster admin in a multi-tenant cluster, I want to provide a cert-manager ClusterIssuer to our users allowing
+self-provisioning of X.509 SVID certificates. The SPIFFE ID (X.509 URI SAN) should identify the namespace
```

Comment:
It'd be nice if `SVID` and `SPIFFE ID` were links to things.

Reply:
Thanks @jsoref! Updated with links.